Refactor common FocusController code path to be site-isolation aware

beidson opened the pull request:
#### 5ae55be289fbe0881e10b5d1c5f3a35b2474583a
<pre>
Refactor common FocusController code path to be site-isolation aware
<a href="https://rdar.apple.com/154258045">rdar://154258045</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294951">https://bugs.webkit.org/show_bug.cgi?id=294951</a>

Reviewed by Ryosuke Niwa.

To prepare for making &quot;advancing focus&quot; to be able to cross frame boundaries with site isolation enabled,
this patch refactors the relevant code path of FocusController to include more context about an attempted
focus operation.

With site isolation disabled, this doesn&apos;t change behavior.

With site isolation enabled, you&apos;ll see the added log statements fire when you try certain operations into
an isolated iframe.

* Source/WebCore/editing/cocoa/AutofillElements.cpp:
(WebCore::nextAutofillableElement):
(WebCore::previousAutofillableElement):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::findFocusableElementDescendingIntoSubframes):
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::findFocusableElementAcrossFocusScope):
(WebCore::FocusController::findFocusableElementWithinScope):
(WebCore::FocusController::nextFocusableElementWithinScope):
(WebCore::FocusController::previousFocusableElementWithinScope):
(WebCore::FocusController::nextFocusableElement):
(WebCore::FocusController::previousFocusableElement):
(WebCore::shouldClearSelectionWhenChangingFocusedElement):
(WebCore::updateFocusCandidateIfNeeded):
* Source/WebCore/page/FocusController.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::nextAssistableElement):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode nextFocusNode]):
(-[DOMNode previousFocusNode]):

Canonical link: <a href="https://commits.webkit.org/296625@main">https://commits.webkit.org/296625@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36029b01ea5b8b076d5513373090be10096f9398

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19183 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114309 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59397 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111061 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29441 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37325 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82912 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112046 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23421 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98264 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63357 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22821 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16406 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/58993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92792 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16449 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117427 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36146 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26719 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91928 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36517 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94528 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91734 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23353 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36644 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14398 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32004 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36043 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41548 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35736 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39075 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37422 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/FocusController.h:
```diff
@@ -50,6 +50,12 @@ class TreeScope;
 struct FocusCandidate;
 
+enum class ContinuedSearchInRemoteFrame : bool { No, Yes };
+struct FindFocusableElementResult {
```
What's the intended values for when the next focusable element in another process?
Are we gonna return nullptr and `ContinuedSearchInRemoteFrame::Yes`?
If so, maybe `ContinuedSearchInRemoteFrame` should be renamed to `ContinueSearchInRemoteFrame`.
It's also inconsistent that this is called "find" focusable element result whereas the enum above is called continued "search" in remote frame.
Maybe rename this to `FocusableElementSearchResult`?

beidson replied:
"ContinuED" <-- as in, past tense -- actually will be accurate, as when the condition is detected is when we'll call out to the remote frame.

(That code has been written, I'm just trying to land in smaller, more manageable patches)

FocusableElementSearchResult is a decent naming suggestion! Will update as I also fight EWS.

lilyspiniolas commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -518,23 +519,30 @@ void FocusController::setFocusedInternal(bool focused)
+    RefPtr element = startingElement;
     while (RefPtr owner = dynamicDowncast<HTMLFrameOwnerElement>(element)) {
+        auto* remoteContentFrame = dynamicDowncast<RemoteFrame>(owner->contentFrame());
+        if (remoteContentFrame) {
```
```suggestion
        if (RefPtr remoteContentFrame = dynamicDowncast<RemoteFrame>(owner->contentFrame()) {
```

lilyspiniolas commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -669,31 +677,35 @@ bool FocusController::advanceFocusInDocumentOrder(FocusDirection direction, Keyb
             // Skip through invokers if they have popovers with focusable contents, and navigate through those contents instead.
-            while (auto* popover = openPopoverForInvoker(candidateInCurrentScope)) {
-                if (auto* candidate = findFocusableElementWithinScope(direction, FocusNavigationScope::scopeOwnedByScopeOwner(*popover), nullptr, event))
+            while (auto* popover = openPopoverForInvoker(candidateInCurrentScope.element.get())) {
```
```suggestion
            while (RefPtr popover = openPopoverForInvoker(candidateInCurrentScope.element.get())) {
```

rniwa commented on Source/WebCore/page/FocusController.h:
```diff
@@ -50,6 +50,12 @@ class TreeScope;
 
 struct FocusCandidate;
 
+enum class ContinuedSearchInRemoteFrame : bool { No, Yes };
```
I still think this should be `ContinueSearchInRemoteFrame` because a function which returns this value didn't continue the search in a remote frame. It's requesting that the caller continue its search in a remote frame as I understand it.

beidson replied:
I know it looks like that with just this patch, but this is just the first patch in a series - In the larger set of completed work, the continued search has already begun.

beidson committed:
Refactor common FocusController code path to be site-isolation aware
rdar://154258045
https://bugs.webkit.org/show_bug.cgi?id=294951

Reviewed by Ryosuke Niwa.

To prepare for making "advancing focus" to be able to cross frame boundaries with site isolation enabled,
this patch refactors the relevant code path of FocusController to include more context about an attempted
focus operation.

With site isolation disabled, this doesn't change behavior.

With site isolation enabled, you'll see the added log statements fire when you try certain operations into
an isolated iframe.

* Source/WebCore/editing/cocoa/AutofillElements.cpp:
(WebCore::nextAutofillableElement):
(WebCore::previousAutofillableElement):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::findFocusableElementDescendingIntoSubframes):
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::findFocusableElementAcrossFocusScope):
(WebCore::FocusController::findFocusableElementWithinScope):
(WebCore::FocusController::nextFocusableElementWithinScope):
(WebCore::FocusController::previousFocusableElementWithinScope):
(WebCore::FocusController::nextFocusableElement):
(WebCore::FocusController::previousFocusableElement):
(WebCore::shouldClearSelectionWhenChangingFocusedElement):
(WebCore::updateFocusCandidateIfNeeded):
* Source/WebCore/page/FocusController.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::nextAssistableElement):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode nextFocusNode]):
(-[DOMNode previousFocusNode]):

Canonical link: https://commits.webkit.org/296625@main
